File: Code/Tools/BuildScripts/angel_build.py
#############################################################################
## Copyright (C) 2008-2009, Shane J. M. Liesegang
## All rights reserved.
## 
## Redistribution and use in source and binary forms, with or without 
## modification, are permitted provided that the following conditions are met:
## 
##     * Redistributions of source code must retain the above copyright 
##       notice, this list of conditions and the following disclaimer.
##     * Redistributions in binary form must reproduce the above copyright 
##       notice, this list of conditions and the following disclaimer in the 
##       documentation and/or other materials provided with the distribution.
##     * Neither the name of the copyright holder nor the names of any 
##       contributors may be used to endorse or promote products derived from 
##       this software without specific prior written permission.
## 
## THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
## AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
## IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
## ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
## LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
## CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
## SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
## INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
## CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
## POSSIBILITY OF SUCH DAMAGE.
##############################################################################

import os
import shutil
import sys
import modulefinder
import compileall
import re
import inspect
import zipfile
import shutil

def recursive_copy(source, dest):
    if (os.path.basename(source)[0] == '.'):
        # print "Skipping", os.path.basename(source)
        return
    
    names = os.listdir(source)
    if not os.path.exists(dest):
        os.makedirs(dest)
    for name in names:
        srcname = os.path.join(source, name)
        dstname = os.path.join(dest, name)
        
        if (os.path.isdir(srcname)):
            recursive_copy(srcname, dstname)
        elif (os.path.basename(srcname)[0] != '.'):
            shutil.copyfile(srcname, dstname)



class PythonPackage:
    def __init__(self, script, remoteLib):
        self.script = script
        self.remoteLib = remoteLib
        
        # Get all the modules that aren't imported by just our scripts, but are
        #  necessary for the interpreter to load up without complaining. 
        self.implicits = [
            'warnings',
            'UserDict',
            'stat',
            'sre_parse',
            'sre_constants',
            'sre_compile',
            'site',
            're',
            'os',
            'ntpath',
            'locale',
            'linecache',
            'copy_reg',
            'codecs',
            ]
        self.mods = []
        self.builtins = []
        self.ignores = ['__main__']
        self.path = []
        
        self.mf = None
        
        
    def setup(self, script):
        pass
    
    def processStartScript(self, script):
        self.path = sys.path[:]
        self.path[0] = os.path.dirname(self.script)
        
        self.mf = modulefinder.ModuleFinder(self.path, 0, ['_angel'])
        
        for mod in self.implicits:
            self.mf.import_hook(mod)
        
        self.mf.run_script(script)
    
    def postProcessResults(self, script):
        alreadies = []
        dict = self.mf.modules
        self.mods = dict.keys()
        for mod in self.mods:
            # clear out any that we're already going to copy over
            if dict[mod].__file__:
                if os.path.dirname(dict[mod].__file__) == os.path.dirname(script):
                    alreadies.append(mod)
            
            # if there's no file (it's built-in to the interpreter), clear it
            if not dict[mod].__file__:
                self.builtins.append(mod)
        
        pre = self.mods
        self.mods = filter(lambda x: x not in self.builtins, self.mods)
        self.mods = filter(lambda x: x not in self.ignores, self.mods)
        self.mods = filter(lambda x: x not in alreadies, self.mods)
    
    
    def copyFiles(self, remoteLib):
        for mod in self.mods:
            fileName = self.mf.modules[mod].__file__
            remoteBase = ''
            loc = os.path.dirname(fileName)
            while (True):
                if (loc in self.path):
                    break
                else:
                    remoteBase = os.path.join(os.path.basename(loc), remoteBase)
                    loc = os.path.dirname(loc)
            
            src = self.mf.modules[mod].__file__
            dest = os.path.join(remoteLib, remoteBase)
            if not os.path.exists(dest):
                os.makedirs(dest)
            shutil.copyfile(src, os.path.join(dest, os.path.basename(src)))
    
    
    def makeItSnappy(self, script, remoteLib):
        compileall.compile_dir(remoteLib)
        compileall.compile_dir(os.path.dirname(script))
    
    
    def makeItTrim(self, remoteLib):
        pyZip = os.path.join(remoteLib, "..", "python25.zip")
        pyz = zipfile.PyZipFile(pyZip, 'w')
        pyz.writepy(remoteLib)
        pyz.close()
        shutil.rmtree(remoteLib)
        pass
    
    def run(self):
        print "Setup..."
        self.setup(self.script)
        print "Process script..."
        self.processStartScript(self.script)
        print "Post-process results..."
        self.postProcessResults(self.script)
        print "Copy files..."
        self.copyFiles(self.remoteLib)
        print "Make it snappy..."
        self.makeItSnappy(self.script, self.remoteLib)
        print "Make it trim..."
        self.makeItTrim(self.remoteLib)
        

def _create_descendant_mapping(class_children):
    """Take a dict of classes mapped to lists of children and return a dict of
    classes mapped to lists of all listed descendants."""
    class_descendants = dict((name, []) for name in class_children.iterkeys())
    for class_name in class_children.iterkeys():
        if not class_descendants[class_name]:
            _resolve_descendant_map(class_name, class_children, class_descendants)
    
    return class_descendants

def _resolve_descendant_map(class_name, class_children, class_descendants):
    """Find all descendants for a class and all of its descendants recursively, and
    insert the descendants into the list in class_descendants."""
    children = class_children[class_name]
    descendants = class_descendants[class_name]
    if not descendants:
        descendants = descendants + children
        for child in children:
            if child in class_descendants:
                if not class_descendants[child]:
                    _resolve_descendant_map(child, class_children, class_descendants)
                for new_descendant in class_descendants[child]:
                    if new_descendant not in descendants:
                        descendants.append(new_descendant)
        class_descendants[class_name] = descendants

def _write_typemaps(filename, class_data):
    inh_file = open(os.path.join(os.getcwd(), filename), 'w')
    inh_file.write("%include <factory.i>\n\n")
    for class_name in class_data.keys():
        inh_file.write("%factory(" + class_name + "*, " + ", ".join(class_data[class_name]) +");\n")
    inh_file.close()
    

def generate_typemaps(files, interface_directory):
    class_data = {}
    class_regex = re.compile(r"\sclass\s(\S*)\s([^\{]*)", re.MULTILINE | re.DOTALL)
    comment_regex = re.compile(r"/\*.*?\*/", re.MULTILINE | re.DOTALL)
    cpp_comment_regex = re.compile(r"//.*")
    for interface_file in files:
        if (os.path.exists(os.path.join(interface_directory, interface_file))):
            file_contents = open(os.path.join(interface_directory, interface_file), 'r').read()
            results = class_regex.findall(file_contents)
            if (results):
                for result in results:
                    derived = result[0]
                    inh = result[1]
                    if inh.find(":") >= 0:
                        # somewhere in here is inheritance
                        
                        # there's likely a more efficient way to do this. but "meh"
                        inh = comment_regex.sub('', inh)
                        inh = cpp_comment_regex.sub('', inh)
                        inh = inh.replace('\n', '')
                        inh = inh.replace(':', '')
                        inh = inh.replace('public', '')
                        inh = inh.replace('protected', '')
                        inh = inh.replace('private', '')
                        
                        for base in inh.split(','):
                            base = base.strip()
                            if base not in class_data.keys():
                                class_data[base] = []
                            class_data[base].append(derived)
                            
    
    descendant_mapping = _create_descendant_mapping(class_data)
    
    chains = {}
    for class_ in descendant_mapping:
        if len(descendant_mapping[class_]) > 0:
            chain = descendant_mapping[class_]
            chain.reverse()
            chains[class_] = chain
            
    out_strings = []
    for class_ in sorted(chains.keys()):
        out_strings.append(class_ + ": " + ", ".join(chains[class_]))
    
    out_string = "\n".join(out_strings)
    
    need_to_generate = False
    out_file_path = os.path.join(os.getcwd(), "build_cache", "typemaps.txt")
    if not os.path.exists(out_file_path):
        if not os.path.exists(os.path.dirname(out_file_path)):
            os.makedirs(os.path.dirname(out_file_path))
        out_file = open(out_file_path, 'w')
        out_file.write(out_string)
        out_file.close()
        need_to_generate = True
    else:
        out_file = open(out_file_path, 'r')
        current_file = out_file.read()
        out_file.close()
        if current_file != out_string:
            out_file = open(out_file_path, 'w')
            out_file.write(out_string)
            out_file.close()
            need_to_generate = True
    
    interface_file = os.path.join(interface_directory, "inheritance.i")
    if not os.path.exists(interface_file):
        need_to_generate = True
    
    if need_to_generate:
        _write_typemaps(interface_file, chains)

File: Code/Tools/BuildScripts/angel_build.lua
------------------------------------------------------------------------------
-- Copyright (C) 2008-2013, Shane Liesegang
-- All rights reserved.
-- 
-- Redistribution and use in source and binary forms, with or without 
-- modification, are permitted provided that the following conditions are met:
-- 
--     * Redistributions of source code must retain the above copyright 
--       notice, this list of conditions and the following disclaimer.
--     * Redistributions in binary form must reproduce the above copyright 
--       notice, this list of conditions and the following disclaimer in the 
--       documentation and/or other materials provided with the distribution.
--     * Neither the name of the copyright holder nor the names of any 
--       contributors may be used to endorse or promote products derived from 
--       this software without specific prior written permission.
-- 
-- THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
-- AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
-- IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
-- ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
-- LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
-- CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
-- SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
-- INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
-- CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
-- ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
-- POSSIBILITY OF SUCH DAMAGE.
------------------------------------------------------------------------------

require "lfs"
require "io"

if (package.path == nil) then
  package.path = ""
end
local mydir = debug.getinfo(1,"S").source:match[[^@?(.*[\/])[^\/]-$]]
if (mydir == nil) then
  mydir = "."
end
package.path = mydir .. "?.lua;" .. package.path
package.path = package.path .. ";" .. mydir .. "lua-lib/penlight-1.0.2/lua/?/init.lua"
package.path = package.path .. ";" .. mydir .. "lua-lib/penlight-1.0.2/lua/?.lua"
package.path = package.path .. ";" .. mydir .. "lua-lib/penlight-1.0.2/lua/pl/?.lua"

require "pl"

-- whether this file is hidden and should thus be ignored
--  (mostly used for skipping .svn directories)
function _isdotfile(filePath)
  if (path.basename(filePath):sub(1,1) == ".") then
    return true
  else
    return false
  end
end

-- because the penlight copy function doesn't preserve case on windows. :-(
function copyfile(from, to)
  if (path.is_windows) then
    os.execute(string.format('copy /Y "%s" "%s" > nul', from, to))
  else
    dir.copyfile(from, to, true)
  end
end

-- this and makedirs blatantly copied from penlight, but without
--  normalizing the case on Windows, because that is dumb
local dirpat
if path.is_windows then
    dirpat = '(.+)\\[^\\]+$'
else
    dirpat = '(.+)/[^/]+$'
end

-- makes a path recursively
--  (there was some good reason at some point to re-implement this here instead
--  of using the penlight version)
function makedirs(p)
  -- windows root drive case
  if(p:find('^%a:$')) then
    return true
  end

  if not path.isdir(p) then
    local subp = p:match(dirpat)
    if not (makedirs(subp)) then
      io.stderr:write("ERROR: Cannot create " .. subp .. "\n")
      os.exit(1)
    end
    return lfs.mkdir(p)
  else
    return true
  end
end

-- recursively copy a directory and its files from src to dst
function recursive_copy(src, dst)
  if (_isdotfile(src)) then
    -- print("Skipping " .. path.basename(src))
    return
  end
  
  if (not path.exists(src)) then
    return
  end

  if (not path.exists(dst)) then
    makedirs(dst)
  end
  
  local names = dir.getfiles(src, "")
  for _, name in pairs(names) do
    if (not _isdotfile(name)) then
      name = path.basename(name)
      local srcname = fulljoin(src, name)
      local dstname = fulljoin(dst, name)
      copyfile(srcname, dstname)
    end
  end
  
  local dirs = dir.getdirectories(src, "")
  for _, dirname in pairs(dirs) do
    dirname = path.basename(dirname)
    local srcname = fulljoin(src, dirname)
    local dstname = fulljoin(dst, dirname)
    recursive_copy(srcname, dstname)
  end
end

-- split a string into a table along a single separator character
function string:split(sep)
  local sep, fields = sep or " ", {}
  local pattern = string.format("([^%s]+)", sep)
  self:gsub(pattern, function(c) fields[#fields+1] = c end)
  return fields
end

--trim leading and following whitespace from a string
function string:trim()
  return self:match("^%s*(.-)%s*$")
end

-- <sigh> the penlight library only joins paths one at a time. :-(
--  (this also corrects for its inability to deal with quoted paths)
function fulljoin(...)
  local arg = {...}
  local isQuoted = false
  -- check to see if any of the paths have quotes in them
  for i, v in ipairs(arg) do
    if (v:find('"')) then
      isQuoted = true
      arg[i] = v:gsub('"', '')
    end
  end

  local for_return = ""
  for i, v in ipairs(arg) do
    for_return = path.join(for_return, v)
  end

  if (isQuoted) then
    for_return = '"' .. for_return .. '"'
  end
  return for_return
end

-- load a file using the passed table as its environment
function loadFileIn(filename, environment)
  if (environment == nil) then
    print "ERROR: Cannot load into nil environment."
  end
  local f, err = loadfile(filename, "bt", environment)
  if (f == nil) then
    print(err)
    return
  end
  return f()
end

-- get the absolute path to the current file
function get_file_location()
  local file_location = debug.getinfo(2,'S').source
  if (file_location:sub(1,1) == "@") then
    file_location = file_location:sub(2, -1)
  end
  return (path.dirname(path.abspath(file_location)))
end

-- correct a given attributions file to claim the used libraries
function correct_attributions(filename, att_dir, disable_devil, disable_fmod)
  local function append_to(file_handle, newfile)
    local newfile_handle = assert(io.open(newfile, "r"))
    local newtext = newfile_handle:read("*all")
    newfile_handle:close()
    file_handle:write(newtext)
  end
  
  local lgpl = false
  
  local f = assert(io.open(filename, "a+"))
  append_to(f, fulljoin(att_dir, "base-angel.txt"))
  if (disable_devil ~= 1) then
    append_to(f, fulljoin(att_dir, "devil-libs.txt"))
    lgpl = true
  end
  if (disable_fmod == 1) then
    append_to(f, fulljoin(att_dir, "ogg-vorbis.txt"))
    if (path.is_windows) then
      append_to(f, fulljoin(att_dir, "openal-soft.txt"))
      lgpl = true
    end
  else
    append_to(f, fulljoin(att_dir, "fmod.txt"))
  end
  
  if (lgpl) then
    append_to(f, fulljoin(att_dir, "lgpl.txt"))
  end
  
  f:close()
end

-- make sure that the installed SWIG is up to date
function check_swig_version(path)
  local f = assert(io.popen(path.." -version", 'r'))
  local s = assert(f:read('*a'))
  f:close()

  version = s:match("\n%s*SWIG Version%s+([^%s]*)")
  vparts = version:split(".")

  -- check that we're at at least 2.0.6
  local old = false
  if (tonumber(vparts[1]) < 2) then
    old = true
  elseif ((tonumber(vparts[2]) < 1) and (tonumber(vparts[3]) < 6)) then
    old = true
  end

  if (old) then
    io.stderr:write("ERROR: swig version "..version.." is too old for Angel; update to at least 2.0.6.\n")
    os.exit(1)
  end
end

-- find a local install of SWIG
function get_swig_path()
  if (path.is_windows) then
    -- we're on windows; use the distributed swig
    return "..\\swigwin-2.0.6\\swig.exe"
  end
  
  local ports_path = "/opt/local/bin/swig"
  if (path.exists(ports_path)) then
    -- this mac user has wisely installed swig from macports
    check_swig_version(ports_path)
    return ports_path
  end

  local brew_path = "/usr/local/Cellar/swig"
  local brew_swig_path = "/usr/local/bin/swig"
  if (path.exists(brew_path)) then
    -- THIS mac user prefers brew to macports. we can support that!
    check_swig_version(brew_swig_path)
    return brew_swig_path
  end

  -- check for other installed swig
  local f = assert(io.popen("which swig", 'r'))
  local s = assert(f:read('*a'))
  f:close()
  if (s == "") then
    io.stderr:write("ERROR: swig not found.\n")
    os.exit(1)
  end
  local other_path = s:gsub("\n" , "")
  check_swig_version(other_path)
  return other_path
end

function find_in_table(haystack, needle)
  for _,v in pairs(haystack) do
    if v == needle then
      return true
    end
  end
  return false
end

-- generate typemap inheritance for SWIG factories
function generate_typemaps(interface_directory, additional_defines)
  local swig = get_swig_path()
  local junkfile = fulljoin(interface_directory, "..", "..", "..", "Tools", "BuildScripts", "build_cache", "swigout.txt")
  local inheritance_file = ""
  if (find_in_table(additional_defines, "INTROGAME")) then
    inheritance_file = fulljoin(interface_directory, "inheritance_intro.i")
  else
    inheritance_file = fulljoin(interface_directory, "inheritance.i")
  end
  if (not path.exists(inheritance_file:gsub('"', ''))) then
    -- need to make sure we have at least a blank file for swig to include 
    --  when it runs to get the type data
    local inheritance_handle = assert(io.open(inheritance_file:gsub('"', ''), "w"))
    inheritance_handle:close()
  end
  local swig_options = ""
  if (additional_defines ~= nil) then
    swig_options = swig_options .. " -D" .. table.concat(additional_defines, " -D")
  end
  swig_options = swig_options .. " -c++ -lua -Werror -debug-typedef -I" .. interface_directory .. " -o " .. junkfile .. " " .. fulljoin(interface_directory, "angel.i")
  
  local exe_string = swig .. swig_options
  local f = assert(io.popen(exe_string, 'r'))
  local s = assert(f:read('*a'))
  f:close()
  
  local separator = "-------------------------------------------------------------"
  local class_data = {}
  local immediate_class_data = {}
  local class_pattern = "Type scope '(%w+)'"
  local inh_patterh = "Inherits from '(%w+)'"
  local splits = s:split(separator)
  for _, v in pairs(splits) do
    local _, _, class = v:find(class_pattern)
    if (class ~= nil) then
      local lines = v:split("\n")
      if (#lines == 1) then
        -- no inheritance here
      else
        immediate_class_data[class] = {}
        for i=1,#lines do -- skip the first line
          local _, _, base = lines[i]:find(inh_patterh)
          if (base ~= nil) then
            if (class_data[base] == nil) then
              class_data[base] = {}
            end
            table.insert(class_data[base], class)
            table.insert(immediate_class_data[class], base)
          end
        end
      end
    end
  end

  local sortf = function(class1, class2)
    local parents1 = immediate_class_data[class1]
    local parents2 = immediate_class_data[class2]
    if (parents1 == nil and parents2 == nil) then
      return false
    end
    if (parents1 ~= nil) then
      for _, p in pairs(parents1) do
        if (p == class2) then
          -- the second class is a parent of the first
          return true
        end
      end
    end
    return false
  end

  local out_strings = {}
  for base, descendants in pairs(class_data) do
    table.sort(class_data[base], sortf)
    table.insert(out_strings, "%factory("..base.."*, "..table.concat(descendants, ", ")..");")
  end
  table.sort(out_strings)

  local out_string = "%include <factory.i>\n\n" .. table.concat(out_strings, "\n")
  
  local out_file = assert(io.open(inheritance_file:gsub('"', ''), "r"))
  local current_file = out_file:read("*a")
  out_file:close()
  if (current_file ~= out_string) then
    out_file = assert(io.open(inheritance_file:gsub('"', ''), "w"))
    out_file:write(out_string)
    out_file:close()
  end
end

